File: backend/app/models/attack.py
"""Attack event ORM model and Pydantic schemas."""

from __future__ import annotations

import enum
from datetime import datetime, timezone
from typing import Optional

from pydantic import BaseModel, ConfigDict
from sqlalchemy import DateTime, Float, Integer, String
from sqlalchemy.orm import Mapped, mapped_column

from app.core.database import Base

# ---------------------------------------------------------------------------
# Enum
# ---------------------------------------------------------------------------


class AttackType(str, enum.Enum):
    DDoS = "DDoS"
    Malware = "Malware"
    Phishing = "Phishing"
    Ransomware = "Ransomware"
    Intrusion = "Intrusion"
    BruteForce = "BruteForce"
    SQLInjection = "SQLInjection"
    XSS = "XSS"
    ZeroDay = "ZeroDay"


# ---------------------------------------------------------------------------
# ORM Model
# ---------------------------------------------------------------------------


class AttackEvent(Base):
    """Persisted record of a single cyber attack event."""

    __tablename__ = "attack_events"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)

    # Source
    source_ip: Mapped[str] = mapped_column(String(45), nullable=False, index=True)
    source_country: Mapped[str] = mapped_column(String(100), nullable=False, index=True)
    source_lat: Mapped[float] = mapped_column(Float, nullable=False)
    source_lng: Mapped[float] = mapped_column(Float, nullable=False)

    # Destination
    dest_ip: Mapped[str] = mapped_column(String(45), nullable=False, index=True)
    dest_country: Mapped[str] = mapped_column(String(100), nullable=False, index=True)
    dest_lat: Mapped[float] = mapped_column(Float, nullable=False)
    dest_lng: Mapped[float] = mapped_column(Float, nullable=False)

    # Attack metadata
    attack_type: Mapped[str] = mapped_column(String(50), nullable=False, index=True)
    severity: Mapped[int] = mapped_column(Integer, nullable=False)
    cluster_id: Mapped[Optional[str]] = mapped_column(
        String(100), nullable=True, index=True
    )

    timestamp: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        default=lambda: datetime.now(timezone.utc),
        index=True,
    )


# ---------------------------------------------------------------------------
# Pydantic schemas
# ---------------------------------------------------------------------------


class AttackEventCreate(BaseModel):
    """Schema for creating a new AttackEvent (inbound)."""

    source_ip: str
    dest_ip: str
    source_country: str
    dest_country: str
    source_lat: float
    source_lng: float
    dest_lat: float
    dest_lng: float
    attack_type: AttackType
    severity: int
    cluster_id: Optional[str] = None
    timestamp: Optional[datetime] = None


class AttackEventResponse(BaseModel):
    """Schema for returning an AttackEvent to API consumers."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    source_ip: str
    dest_ip: str
    source_country: str
    dest_country: str
    source_lat: float
    source_lng: float
    dest_lat: float
    dest_lng: float
    attack_type: str
    severity: int
    cluster_id: Optional[str]
    timestamp: datetime

File: frontend/types/attack.ts
export enum AttackType {
  DDoS = "DDoS",
  Malware = "Malware",
  Phishing = "Phishing",
  Ransomware = "Ransomware",
  Intrusion = "Intrusion",
  BruteForce = "BruteForce",
  SQLInjection = "SQLInjection",
  XSS = "XSS",
  ZeroDay = "ZeroDay",
}

export interface AttackEvent {
  id: string;
  source_ip: string;
  dest_ip: string;
  source_country: string;
  dest_country: string;
  source_lat: number;
  source_lng: number;
  dest_lat: number;
  dest_lng: number;
  attack_type: AttackType;
  severity: number;
  timestamp: string;
  cluster_id?: string;
}

export interface Stats {
  events_per_second: number;
  rolling_avg: number;
  is_anomaly: boolean;
  anomaly_score: number;
  top_attackers: Array<{ ip: string; country: string; count: number }>;
  top_targets: Array<{ country: string; count: number }>;
  attack_type_stats: Record<string, number>;
  total_events: number;
}

export type WebSocketMessage =
  | { type: "attack"; data: AttackEvent }
  | { type: "stats"; data: Stats }
  | { type: "anomaly"; data: { message: string; score: number } }
  | { type: "history"; data: AttackEvent[] }
  | { type: "replay_started"; data: { total: number } }
  | { type: "replay_stopped"; data: Record<string, never> }
  | { type: "replay_seek"; position: number }
  | { type: "alert"; data: { rule_id: number; rule_name: string; message: string; fired_at: number } };
